src/toolbars/AudioSetupToolBar.cpp

```diff
@@ -0,0 +1,856 @@
+   auto& device = isInput ? mInput : mOutput;
+
+   auto host = AudioIOHost.Read();
+   const std::vector<DeviceSourceMap>& maps = isInput ? DeviceManager::Instance()->GetInputDeviceMaps()
+      : DeviceManager::Instance()->GetOutputDeviceMaps();
+
+   auto item = device->FindChildItem(deviceId);
+   if (item) {
+      // Update cache with the chosen device
+      item->Check();
+      wxString newDevice = item->GetItemLabelText();
+
+      for (size_t i = 0; i < maps.size(); ++i) {
+         wxString name = MakeDeviceSourceString(&maps[i]);
+         if (name == newDevice && maps[i].hostString == host) {
+            newIndex = i;
+         }
+      }
+   }
+
+   if (newIndex < 0) {
+      wxLogDebug(wxT("AudioSetupToolBar::OnMenu(): couldn't find device indices"));
+      return;
+   }
+
+   SetDevices(isInput ? &maps[newIndex] : nullptr,
+              isInput ? nullptr : &maps[newIndex]);
+}
+
+void AudioSetupToolBar::OnMenu(wxCommandEvent& event)
+{
+   int id = event.GetId();
+   bool audioSettingsChosen = false;
+
+   if ((id >= kHost) && (id < kInput)) {
+      ChangeHost(id);
+   }
+   else if ((id >= kInputChannels) && (id < kOutput)) {
+      if (auto item = mInputChannels->FindChildItem(id)) {
+         // Update cache with selected number of input channels
+         item->Check();
+         AudioIORecordChannels.Write(id - kInputChannels + 1);
+      }
+   }
+   else if ((id >= kInput) && (id < kInputChannels)) {
+      ChangeDevice(id, true);
+   }
+   else if ((id >= kOutput) && (id < kAudioSettings)) {
+      ChangeDevice(id, false);
+   }
+   else if (id == kAudioSettings) {
+      audioSettingsChosen = true;
+   }
+
+   auto gAudioIO = AudioIOBase::Get();
+   if (gAudioIO) {
+      // We cannot have gotten here if gAudioIO->IsAudioTokenActive(),
+      // per the setting of AudioIONotBusyFlag and AudioIOBusyFlag in
+      // AudacityProject::GetUpdateFlags().
+      // However, we can have an invalid audio token (so IsAudioTokenActive()
+      // is false), but be monitoring.
+      // If monitoring, have to stop the stream, so HandleDeviceChange() can work.
+      // We could disable the Preferences command while monitoring, i.e.,
+      // set AudioIONotBusyFlag/AudioIOBusyFlag according to monitoring, as well.
+      // Instead allow it because unlike recording, for example, monitoring
+      // is not clearly something that should prohibit changing device.
+      // TODO: We *could* be smarter in this method and call HandleDeviceChange()
+      // only when the device choices actually changed. True of lots of prefs!
+      // As is, we always stop monitoring before handling the device change.
+      if (gAudioIO->IsMonitoring())
+      {
+         gAudioIO->StopStream();
+         while (gAudioIO->IsBusy()) {
+            using namespace std::chrono;
+            std::this_thread::sleep_for(100ms);
+         }
+      }
+
+      if (audioSettingsChosen) {
+         if (auto project = ::GetActiveProject().lock()) {
```

Comment:
New uses of `GetActiveProject` should be avoided!  It dependes on a global variable to identify the project.

But this toolbar was constructed with a reference to the project, which was passed to the bass class ToolBar.

Give ToolBar a const accessor to return AudacityProject& and use that.

Reply:
Actually, it's a projecte member, so just use `mProject` directly.

Reply:
Done.